devnet-sdk/devstack/sysgo/sync_test.go

```diff
@@ -219,3 +219,126 @@ func TestL2CLSyncP2P(gt *testing.T) {
+			blockA, err := elA.EthClient().BlockRefByLabel(ctx, label)
+			require.NoError(t, err)
+			blockA2, err := elA2.EthClient().BlockRefByLabel(ctx, label)
+			require.NoError(t, err)
+			cancel()
+			logger.Info("chain A", "blockNum", blockA.Number, "block", blockA)
+			logger.Info("chain A2", "blockNum", blockA2.Number, "block", blockA2)
+			return blockA, blockA2
+		}
+
+		queryCL := func() (*eth.SyncStatus, *eth.SyncStatus) {
+			ctx, cancel := context.WithTimeout(context.Background(), time.Second*10)
+			syncA, err := clA.RollupAPI().SyncStatus(ctx)
+			require.NoError(t, err)
+			syncA2, err := clA2.RollupAPI().SyncStatus(ctx)
+			require.NoError(t, err)
+			cancel()
+			logger.Info("chain A", "sync", syncA)
+			logger.Info("chain A2", "sync", syncA2)
+			return syncA, syncA2
+		}
+
+		querySupervisor := func(chainID eth.ChainID) eth.BlockID {
+			ctx, cancel := context.WithTimeout(context.Background(), time.Second*10)
+			view, err := supervisor.QueryAPI().LocalUnsafe(ctx, chainID)
+			require.NoError(t, err)
+			cancel()
+			return view
+		}
+
+		targetBlockNum1 := uint64(10)
+		logger.Info("wait until reaching target block", "blockNum", targetBlockNum1)
+		require.Eventually(t, func() bool {
+			blockA, blockA2 := queryEL(eth.Unsafe)
+			return blockA.Number >= targetBlockNum1 && blockA2.Number >= targetBlockNum1
+		}, 30*time.Second, waitTime)
+
+		logger.Info("disconnect p2p between L2CLs")
+		DisconnectL2CLP2P(ids.L2ACL, ids.L2A2CL)(orch)
+
+		// verifier lost its P2P connection with sequencer, and will advance its unsafe head by reading L1 but not by P2P
+		_, prevblockA2 := queryEL(eth.Unsafe)
+		targetBlockNum2 := prevblockA2.Number + 5
+		logger.Info("make sure verifier advances safe head by reading L1", "blockNum", targetBlockNum2)
+		require.Eventually(t, func() bool {
+			_, syncA2 := queryCL()
+			// unsafe head and safe head both advanced from last observed unsafe head
+			return syncA2.SafeL2.Number == syncA2.UnsafeL2.Number && syncA2.SafeL2.Number > targetBlockNum2
+		}, 60*time.Second, waitTime)
+
+		logger.Info("verifier heads will lag compared from sequencer heads and supervisor view")
+		require.Never(t, func() bool {
+			syncA, syncA2 := queryCL()
+			chainAView := querySupervisor(elA2.ChainID())
+			// unsafe head will always lagged
+			check := syncA.UnsafeL2.Number > syncA2.UnsafeL2.Number
+			// safe head may be matched or lagged
+			check = check && syncA.SafeL2.Number >= syncA2.SafeL2.Number
+			// unsafe head may be matched or lagged compared to supervisor unsafe head view for chain A
+			check = check && chainAView.Number >= syncA2.UnsafeL2.Number
+			logger.Info("unsafe head sync status", "sequencer", syncA.UnsafeL2.Number, "supervisor", chainAView.Number, "verifier", syncA2.UnsafeL2.Number)
+			return !check
+		}, 15*time.Second, waitTime)
+
+		logger.Info("explicit reconnection of L2CL P2P between sequencer and verifier")
+		WithL2CLP2PConnection(ids.L2ACL, ids.L2A2CL)(orch)
```

Comment:
Nit (this can be part of the same follow-up PR that generalizes this test to not be in the sysgo package anymore): it would be nice to have DSL functions for the connecting/disconnecting of peers. We shouldn't apply Orchestrator options after the system frontend to the test is already up and running.

Reply:
Opened an issue to track: https://github.com/ethereum-optimism/optimism/issues/15700